torch_xla/experimental/stablehlo_saved_model.py

```diff
@@ -0,0 +1,110 @@
+import copy
+import shutil
+import os
+import re
+import numpy as np
+import torch
+from torch import nn
+import torch_xla
+from torch_xla.core import xla_model as xm
+import tensorflow as tf
+import torchvision
+import torch._dynamo as torchdynamo
+
+from tensorflow.compiler.tf2xla.python import xla as tfxla  # type: ignore[import]
+
+from typing import Tuple, Type, Callable
+
+import sys
+
+
+class SHLOModel:
+
+  def __init__(self, model_bytecode, pos_to_orig_pos, pos_to_param,
+               sample_output):
+    self.model_bytecode = model_bytecode
+    self.pos_to_orig_pos = pos_to_orig_pos
+    self.pos_to_param = pos_to_param
+    self._total_number = len(pos_to_orig_pos) + len(pos_to_param)
+    output_tf = tf.convert_to_tensor(sample_output)
+    self.tout = (output_tf.dtype,)
+    self.sout = (tuple(output_tf.shape),)
+
+  def __call__(self, *args):
+
+    call_args = []
+    for i in range(self._total_number):
+      if i in self.pos_to_orig_pos:
+        call_args.append(args[self.pos_to_orig_pos[i]])
+      else:
+        call_args.append(self.pos_to_param[i])
+
+    return tfxla.call_module(
+        tuple(call_args),
+        version=5,
+        Tout=self.tout,  # dtype information
+        Sout=self.sout,  # Shape information
+        function_list=[],
+        platforms=('CPU',),
+        module=self.model_bytecode,
+    )[0]
+
+
+def export_torch_model(model: torch.nn.Module,
+                       sample_inputs: Tuple[torch.Tensor]):
+  """Convert model into a callable backed by StableHLO.
+    
+    Args:
+        model: torch.nn.Module - a pytorch model
+        sample_inputs: Tuple[torch.Tensor] - The input to this model
+
+    This function will return a callable backed by StableHLO such that,
+
+    model(*sample_inputs) ==  export_torch_model(model, sample_inputs)(*sample_inputs)
+    (up to numerics)
+
+    In other words, returned callable have the same calling convention of the input model, and
+    on the sample input, or inputs sufficiently similar* to sample input,  
+    it is will to return same result as the original model.
+
+    * sufficiently similar input because this function will use tracing to extract the model operations
+      so it might specialize on the shapes of the sample input.
+    
+    For now, model has to only take Tensors as input and has to return 1 tensor as output.
+
+    """
+  xm.mark_step()
+  model = copy.deepcopy(model)
+  sample_inputs = copy.deepcopy(sample_inputs)
+
+  model.to(device='cpu')
+  sample_outputs = model(*sample_inputs)
+
+  device = xm.xla_device()
+  model.to(device=device)
+  sample_inputs_lazy = tuple(map(lambda x: x.to(device=device), sample_inputs))
+  input_ids = {
+      torch_xla._XLAC._xla_get_tensor_id(tensor): i
+      for i, tensor in enumerate(sample_inputs_lazy)
+  }
+  output = model(*sample_inputs_lazy)
```

Comment:
You want to call `clear_pending_irs` similar to https://github.com/pytorch/xla/blob/b3f4c7bc1e9f812069770c83e38e5bd2dc20aef6/torch_xla/core/dynamo_bridge.py#L299-L305

otherwise you will leave a pending graph on device which might mess up the future execution and wasting compute.

Reply:
done

Reply:
I think you can also get `output_shape` and `output_dtype` by doing `output.dtype` and `otuput.size()` here

Reply:
would that materialize `output? and thus lose the LazyIR stored in it?

Reply:
nope, shape and dtype are generated during tracing, it won't trigger execution.

Reply:
Took a another look at this pr, once we fixed this comment, this pr lgtm

Reply:
done